### PR TITLE
Addition of multiple Update rules (++)

### PR DESCRIPTION
**Pull request by Schokobecher**
Added rules for WinSCP (this time without conflicts, yay!), ProcessHacker, Everything (Search) - and also while I'm at it mpc-hc, deluge, sharex, lockhunter, obs-studio, colemak, handbrake and WinDirStat.

SumatraPDF is really annoying, the main site uses JS to display the version and download links, the changelog is structured in such a way that I can't figure out a good css selector. So here we go with 43 matches and counting.

### CODE REVIEW

**Comment by Schokobecher** [comment]
@guiweber That's it for now. Time for a break. Merry X-Mas and such!

**Comment by guiweber** [comment]
Thanks for all these additions. I'm currently out of the country with very limited Internet access,  so I'll review this pull request and anything else you have submitted when I come back.

A good holiday to you as well!

**Comment by Schokobecher** [comment]
Don't worry about it.
In [5a3106f](https://github.com/guiweber/just-install-updater/pull/8/commits/5a3106ffc55595db348e786e8e44c7824774b4c6) I added another replacement scheme to `jiup-lib.js`:

>   final = final.replace(/{{\.version#}}/g, version.replace(/\./g, ""));


because `dependency-walker` doesn't use any of the given ones for it's files. 
If you are unhappy with the usage of the `#` symbol, I'm open to suggestions (couldn't think of any that wouldn't break JS or are used otherwise)

File: update-rules.json [hunk 1]
```diff
@@ -930,5 +930,30 @@
       }
     },
 	"version_extractor": "([.\\d]+[.]\\d{1,5})"
+  },
+     "handbrake": {
+     "url": "https://handbrake.fr/downloads.php",
+     "updater": {
+      "x86":{
+        "rule_type": "css-link",
+        "selector": "[href$='i686-Win_GUI-1.exe']"
```
Comment (guiweber) [inline update-rules.json]: It looks like  the naming convention changes from version to version. The "-1" is not always included in the file name, so a more generic selector will be needed.

File: update-rules.json [hunk 2]
```diff
@@ -829,5 +828,286 @@
       }
     },
     "version_extractor": "^[0-9]{2}"
+  },
+    "winscp": {
+     "url": "https://winscp.net/eng/download.php",
+     "updater": {
+      "x86":{
+        "rule_type": "css-html-version",
+        "selector": "#content > h3:nth-child(8)",
```
Comment (guiweber) [inline update-rules.json]: using "nth-child" is very risky as most changes to the page will break the rule. Better use regexp to get a surer match on the page, such as a section title on the page with the version number (e.g. WinSCP 5.9.3)

File: update-rules.json [hunk 3]
```diff
@@ -829,5 +828,286 @@
...
+      },
+      "x86_64":{
+        "rule_type": "css-link",
+        "selector": "a[href$='.x64.exe']"
+      }
+    },
+    "version_extractor": "([.\\d]+[.]\\d{1,5})"
+  },
+   "deluge": {
+    "url": "http://dev.deluge-torrent.org/wiki/Download",
```
Comment (guiweber) [inline update-rules.json]: for Deluge have you tried getting the link directly from the download page (http://download.deluge-torrent.org/windows/?C=M;O=D) where all the links are located? Their file naming convention doesn't seem very constant so I'm a bit nervous about getting the version number only.

File: update-rules.json [hunk 4]
```diff
@@ -829,5 +828,286 @@
...
+      }
+    },
+	"version_extractor": "([.\\d]+[.]\\d{1,5})"
+  },
+   "handbrake": {
+    "url": "https://handbrake.fr/downloads.php",
+    "updater": {
+      "x86":{
+        "rule_type": "css-link",
+        "selector": "[href$='i686-Win_GUI.exe']"
```
Comment (guiweber) [inline update-rules.json]: It looks like the naming convention changes from version to version. A suffix (e.g. "-1") is sometimes included in the file name, so a more generic selector would be better in order to get matches when a suffix is included.

File: update-rules.json [hunk 5]
```diff
@@ -829,5 +828,286 @@
...
+      }
+    },
+	"version_extractor": "([.\\d]+[.]\\d{1,5})"
+  },
+   "kodi": {
+    "url": "https://kodi.tv/download",
+    "updater": {
+      "x86":{
+        "rule_type": "css-link",
+        "selector": "[href$='Jarvis.exe']"
```
Comment (guiweber) [inline update-rules.json]: Jarvis seems to be the name of the current release. The next release will be named "Krypton" so it's not a good idea to use this as a selector.

File: update-rules.json [hunk 6]
```diff
@@ -829,5 +828,286 @@
...
+      }
+    },
+    "version_extractor": "([.\\d]+[.]\\d){1,5}"
+  },
+    "emacs": {
+    "url": "http://ftp.gnu.org/gnu/emacs/windows/",
+    "updater": {
+      "x86":{
+        "rule_type": "css-link",
+        "selector": "[href$='i686-w64-mingw32.zip']"
```
Comment (guiweber) [inline update-rules.json]: I'm not sure I understand their naming convention. It seems to vary wildly from a release to the next. Do you think this rule will be able to match future releases?

File: update-rules.json (hunk 3 above)
Comment (Schokobecher) [inline update-rules.json]: To be honest, the last release was `2016-07-20 22:47` I don't think there is a much better way, if they can't decide on `2.7.0` oder `2.7` for the python version.. only time will tell - If you really want I can experiment but it's hard to tell what they gonna do next

File: update-rules.json (hunk 4 above)
Comment (Schokobecher) [inline update-rules.json]: I thought it was an error on their end (like having duplicate files on a disk) but check [3616771](https://github.com/guiweber/just-install-updater/pull/8/commits/36167719fbd53417c35943e390e2428933c0c951) for  a more permissive version

File: update-rules.json (hunk 6 above)
Comment (Schokobecher) [inline update-rules.json]: Their older versions had a different scheme to it, all we can do is hope for the best - at least for now. The last 2 version have the current scheme

File: update-rules.json (hunk 3 above)
Comment (guiweber) [inline update-rules.json]: A filter could be used to check that it at least includes "py2.7". I'll merge as-is and see if I can tweak it later.

File: update-rules.json (hunk 4 above)
Comment (guiweber) [inline update-rules.json]: Thanks! I think it's more like a hotfix or minor patch. Many devs add such suffixes for minor patches (e.g. Deluge has just the same: 13.12 and 13.12-1 released a few days later).

File: update-rules.json (hunk 3 above)
Comment (Schokobecher) [inline update-rules.json]: Please check [2925e34](https://github.com/guiweber/just-install-updater/pull/8/commits/2925e3461a844aefee2646b713f1ff59b96b2a07) I find it really messy (mostly had to work against mod_autoindex) but it works. I would prefer if you can come up with something nicer 👍

File: update-rules.json (hunk 5 above)
Comment (Schokobecher) [inline update-rules.json]: I'm currently working on an advanced-rule but my js skills are very limited. Steps would be:

1. Parse Page (https://kodi.tv/download/)
2. Grab Version + Code name
3. Combine into kodi-VERSION-CODENAME.exe
4. Build Link
5. Return Link

I don't think there is any other way?

Comment (guiweber) [inline update-rules.json]: without using an advanced rule, we could just check all the links to an executable file (`"selector": "[href$='.exe']"`) and then filter out anything that has a special tag after the release name, such as "alpha", "beta" or "rc". A filter like: `"-[a-zA-Z]*\\.exe$"` might just do the trick.



Also, maybe getting the links there could be easier: http://mirrors.kodi.tv/releases/win32/

Comment (Schokobecher) [inline update-rules.json]: Ah yes, that could work, I already created an advanced rule [dcb3a8f](https://github.com/guiweber/just-install-updater/pull/8/commits/dcb3a8f0785ab54d387640aeec060d3ab6a2e964) but you don't have to merge it - I will try your idea.

Comment (Schokobecher) [inline update-rules.json]: Ok final commit on this one.
[a4f2a85](https://github.com/guiweber/just-install-updater/pull/8/commits/a4f2a85a91fbda31cddb62cf79f89845d119e49f) removes the advanced rule and replaces it with your idea

Comment (guiweber) [inline update-rules.json]: Could you just remove advanced_rules/kodi.js ? I'll merge your changes once the file is removed. Thanks!

Comment (Schokobecher) [inline update-rules.json]: File has been removed. [169208a](https://github.com/guiweber/just-install-updater/pull/8/commits/169208aa40bd9f72beae694fe81ab6ae5b3f171a)

**Comment by guiweber** [comment]
Thanks again for all the additions!

**Pull request merged**